recipes/LibriSpeech/quantization/train_kmeans.py

```diff
@@ -0,0 +1,258 @@
+
+"""Recipe for training a K-means quantizer on features from an SSL model.
+
+To run this recipe:
+> python train_kmeans.py hparams/<config>.yaml
+
+Authors
+ * Luca Della Libera 2024
+"""
+
+import sys
+
+import torch
+import torchaudio
+from hyperpyyaml import load_hyperpyyaml
+
+import speechbrain as sb
+from speechbrain.utils.distributed import if_main_process
+
+
+class Quantization(sb.Brain):
+    def compute_forward(self, batch, stage):
+        """Forward pass."""
+        batch = batch.to(self.device)
+        sig, lens = batch.sig  # [B, T]
+
+        # Extract features
+        with torch.no_grad():
+            self.modules.ssl_model.eval()
+            feats = self.modules.ssl_model(sig, lens)  # [K, B, N, H]
+            feats = feats[self.hparams.layer_id]  # [B, N, H]
+
+        return feats
+
+    def compute_objectives(self, predictions, batch, stage):
+        """Computes the objectives."""
+        feats = predictions  # [B, N, H]
+
+        # Accumulate features
+        if stage == sb.Stage.TRAIN:
+            feats = feats.flatten(end_dim=-2)  # [BN, H]
+            self.feats.append(feats)
+            self.curr_batch_size += len(feats)
+            if self.curr_batch_size < self.hparams.kmeans_batch_size:
+                # Leave average loss unchanged
+                loss = torch.tensor(self.avg_train_loss)
+                # Workaround to keep consistency with standard supervised training
+                loss.requires_grad_()
+                return loss
+            self.feats = torch.cat(self.feats)
+            feats = self.feats[: self.hparams.kmeans_batch_size]
+            self.feats = [self.feats[self.hparams.kmeans_batch_size :]]
+            self.curr_batch_size = len(self.feats[0])
+
+        if stage == sb.Stage.TRAIN:
+            # Retrieve current centroids
+            old_cluster_centers = self.hparams.quantizer.cluster_centers
+
+            # Partial fit on current batch
+            self.hparams.quantizer.partial_fit(feats)
+
+            # For K-means the training loss is the drift between current centroids and old centroids
+            # If close to 0, it means that the training has converged
+            curr_cluster_centers = self.hparams.quantizer.cluster_centers
+            loss = (curr_cluster_centers - old_cluster_centers).norm()
+
+            # Workaround to keep consistency with standard supervised training
+            loss.requires_grad_()
+            self.optimizer_step += 1
+            assert self.optimizer_step == self.modules.quantizer.n_steps, (
+                self.optimizer_step,
+                self.modules.quantizer.n_steps,
+            )
+        else:
+            # For K-means the validation/test loss is the inertia
+            # The lower the inertia, the better should be the clustering
+            # It is useful to monitor progress across epochs
+            # However, when saving checkpoints we always keep the last one (i.e. max_keys=["epoch"])
```

Comment:
I don't suppose anyone has tested if clusters with lower inertia provide better quantization, i.e. better performance on downstream tasks...

Reply:
Not that I'm aware of